### PR TITLE
add tool to evacuate LBaaSv2 loadbalancers from a node

### PR DESCRIPTION
**Pull request by rhafer**
The new tool is primary needed for the upgrade case. Where we need to free up nodes before upgrading them. But is might also be useful for other maintenance tasks sometime.

Unfortunately neutron-lbaas doesn't provide API to reassign loadbalancers, so we need to operated on the database directly and force agent restarts.

### CODE REVIEW

**Comment by rhafer** [comment]
[S57P10 (8) LBaaS migrations during non-disruptive upgrade](https://trello.com/c/usG7xO7K/364-s57p10-8-lbaas-migrations-during-non-disruptive-upgrade)

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [inline files/default/neutron-evacuate-lbaasv2-agent.py]: Hm, this is now also preventing the namespace removal (which needs to happen anyway)

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [inline files/default/neutron-evacuate-lbaasv2-agent.py:147]: Are the IDs of load balancers printed here just some integers, or the UUIDs?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [inline files/default/neutron-evacuate-lbaasv2-agent.py]: Is there any reason why we wouldn't want to restart the agent? Is this for the upgrade case? If yes, maybe we should make restart the default action, so the script is useful by default outside of upgrade.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [inline files/default/neutron-evacuate-lbaasv2-agent.py:175]: Not that I mind the code, but out of curiosity: in which case would we want to not use crm? If we have multiple agents, then we're in HA mode, and therefore we have crm.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [inline files/default/neutron-evacuate-lbaasv2-agent.py]: Can we rename to namespaces instead of namesspaces?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [inline files/default/neutron-evacuate-lbaasv2-agent.py]: Am I guessing right that we only want this when we don't want to restart the agent? My question really is: do we need the two parameters?

**Review by vuntz — COMMENTED** [review]
Generally looks good to me, some minor comments/questions.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:147]: It's the UUIDs

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py]: Yes this is a upgrade specific feature. (The source agents are down after the first network node is upgraded)

But the default for "cfg.CONF.source_agent_restart" is acatually true. When needed we call the tool with "--nosource_agent_restart".

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:175]: I agree that switch is probably useless for SUSE Cloud. Though in theory you don't need a cluster for running multiple lbaas agents.

I used it during development when I had all clusters put into maintenance mode permanently.

Shall I remove it? Or at least switch the default?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py]: Hm, I guess you're right.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py]: Sure

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py]: @vuntz  Hm, actually there are some cases where we don't want to restart the agent or remove the namespaces. That is when we know that the source node is offline. But actually the current code doesn't handle that case very well. (It would still try to open an ssh connection if both settings are turned off :( )

Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py]: Actually there is still a bug which prevents " cfg.CONF.delete_namespaces" from working correctly. I though I already addressed that :/

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (vuntz) [reply files/default/neutron-evacuate-lbaasv2-agent.py:175]: It's fine to keep it that way, I was just curious.

**Comment by rhafer** [comment]
> Actually there is still a bug which prevents " cfg.CONF.delete_namespaces" from working correctly. I though I already addressed that :/

This should be fixed now.

**Review by vuntz — APPROVED** [review]
Giving a +1; I didn't test this and wasn't super careful in the review, but let's unblock this.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py:1103]: Adding the `namespace` parameter to all of this following chunk suggests a possible [Data Envy code smell](http://wiki.c2.com/?DataEnvy).  This is strengthened by the presence of a bunch of `netns_*` instance attributes. 
 Would it make sense to extract a `Namespace` class?

**Review by aspiers — COMMENTED** [review]
Looks fine at a quick glance, but I find myself wondering WWMD (What Would Mate Do) about that namespace-specific code, since he is rapidly becoming the king of beautifully clean and elegant code :-)

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (AbelNavarro) [inline files/default/neutron-evacuate-lbaasv2-agent.py:37]: this could fail should `neutron-ha-tool` not being present in the system. In our environments it's unlikely we'll find that case so I don't know if we need to control that.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (AbelNavarro) [inline files/default/neutron-evacuate-lbaasv2-agent.py:148]: `available_destination_agents()` can return an empty list. I would return an error and exit the script.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (AbelNavarro) [inline files/default/neutron-evacuate-lbaasv2-agent.py:115]: An improvement for this would be to split the reassigned load balancers among the target agents. I.e. three agents (A, B, C) with a number of LB (100, 1, 2). In this case if we want to reassign A it will end up in B with 101 LB. Better would be B(51) and C(52).

This is border case since it's unlikely all LBs end up in one agent, but if we start reassigning LBs with this script we could reach that point. In any case, this is an improvement for the future.

Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:115]: > An improvement for this would be to split the reassigned load balancers among the target agents. I.e. three agents (A, B, C) with a number of LB (100, 1, 2). In this case if we want to reassign A it will end up in B with 101 LB. Better would be B(51) and C(52).

Actually the code does distribute the loadbalancers equally among the available agents. It updates the load of the agent after each assignment (line 125) and get the new minimum for each loadbalancer

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:37]: It will fail like this:

```
Traceback (most recent call last):
  File "/usr/bin/neutron-evacuate-lbaasv2-agent", line 37, in <module>
    '', os.path.join(dirname, '/usr/bin/neutron-ha-tool'))
IOError: [Errno 2] No such file or directory
```

It's not the most user friendly kind of failure, but somehow I think it's good enough for a tool like this.

File: files/default/neutron-ha-tool.py
Comment (rhafer) [reply files/default/neutron-ha-tool.py:1103]: Hm, I don't smell anything ;)

On a more serious note I don't quite see how the introduction of a separate Namespace class would improve the current code. I am not saying that the current implementation is perfect, but it might be good enough (TM) already.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:148]: @abel-navarro Yeah, true. I added that.

**Review by AbelNavarro — APPROVED** [review]

File: files/default/neutron-ha-tool.py
Comment (aspiers) [reply files/default/neutron-ha-tool.py:1103]: > On a more serious note I don't quite see how the introduction of a separate Namespace class would improve the current code.

You mean you understand my point but you disagree? If so, why? If not, I can try to explain further.

@matelakat What do you think?

> I am not saying that the current implementation is perfect, but it might be good enough (TM) already.

Yes it's probably good enough (in this respect at least), but that's not a good reason to avoid a small and easy change if it makes it even better.

Comment (aspiers) [reply files/default/neutron-ha-tool.py:1103]: For example, look at https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/27/files#diff-937ed0b9cba9efd2521c8fd1c9e8416fR402. @matelakat didn't *have* implement the `AgentList` class - he could have just used a `list` and put `calculate_maximal_router_count_difference` and related helper methods into one of the existing classes. But by choosing to separate these concerns into a new class rather than bloat an existing one, he ended up writing code which was beautifully clean, easy to read and maintain (and even unit-test, if that was necessary), for very little extra effort. Maybe I am missing a reason why this couldn't be done with a `Namespace` class?

**Comment by rhafer** [comment]
> Yes it's probably good enough (in this respect at least), but that's not a good reason to avoid a small and easy change if it makes it even better.

... unless there are more pressing issues to address :(

**Comment by aspiers** [comment]
I've merged #27 now.  Even though I checked manually that a merge of both this and #27 still passes the test suite, it might be worth ensuring that via Travis by rebasing.

File: files/default/neutron-ha-tool.py
Comment (rhafer) [reply files/default/neutron-ha-tool.py:1103]: > Yes it's probably good enough (in this respect at least), but that's not a good reason to avoid a small and easy change if it makes it even better.

It probably becomes a good reason once we take other constraints into consideration. Like e.g. time pressure.

Comment (aspiers) [reply files/default/neutron-ha-tool.py:1103]: If that's the only reason then I'd say it's a false economy, since technical debt always increases in cost over time.  And @matelakat is currently blazing the trail and showing us all that it's perfectly possible to write really nice code under time constraints, which is quite inspiring :-)  But ultimately it's your call.  I don't think I want to give a -1 or a +1 right now.  Everything else here looks good to me though.

File: files/default/test-neutron-evacuate-lbaasv2-agent.py
Comment (matelakat) [inline files/default/test-neutron-evacuate-lbaasv2-agent.py:36]: You have a typo here (nit)

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (matelakat) [inline files/default/neutron-evacuate-lbaasv2-agent.py:153]: Why are we using parentheses here?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (matelakat) [inline files/default/neutron-evacuate-lbaasv2-agent.py:146]: saying `if not load_balancers` and exiting would dedent the following lines rendering them more readable.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (matelakat) [inline files/default/neutron-evacuate-lbaasv2-agent.py:172]: we can pass the `cfg.CONF.use_crm` to `RemoteLbaasV2Cleanup` and it can decide what mechanism to use to restart the agent? One less conditional here?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (matelakat) [inline files/default/neutron-evacuate-lbaasv2-agent.py:130]: Does this mean that it will load `/etc/neutron.conf` ?

File: files/default/neutron-ha-tool.py
Comment (matelakat) [inline files/default/neutron-ha-tool.py:1167]: we could have avoided the interface change by overriding the constructor and setting the router id as a member variable.

**Review by matelakat — COMMENTED** [review]
I took a look at the change, left some comments. I think what we could do is to separate concerns a bit. One idea is to have one class that is only responsible for handling the ssh related functions. Something that can execute commands on systems. That would basically hide the fact that we are talking to a remote system. It would have a pretty simple interface, like one function, `run_command` or something like that. On top of that, we could have another class that represents the things that we do with a host: list namespaces, delete namespaces, kill pids. At the moment these responsibilities are not separated, and classes do not hide details - they both know about ssh, they both connect. `RemoteNodeCleanup` is used as a superclass for `RemoteLbaasV2Cleanup`, using it's `_ssh_connect` and `_ssh_client` methods - meaning low level connection, and also to execute arbitrary code on the host with the `_simple_ssh_command`. Let's have a chat, maybe we would need to merge it and then improve later, or we could give it a try to fix it now - maybe a pair programming session?

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:130]: ```/etc/neutron/neutron.conf``` and everything in  ```/etc/neutron/neutron.conf.d/```. Or whatever is specified on the command line. This is to get the database connection details and agent timeouts. 

(Everything can of course also be overridden via separate command line options.

File: files/default/neutron-evacuate-lbaasv2-agent.py
Comment (rhafer) [reply files/default/neutron-evacuate-lbaasv2-agent.py:153]: Why not? :) Yeah, I will remove those.

**Review by matelakat — APPROVED** [review]
We had a discussion with @rhafer and we booked a separate refactor pair programming session where we try to improve the design of the code.

**Comment by matelakat** [comment]
One other note: we will need to document the usage of this tool - what are its requirements, what environment variables it assumes, what configuration files it uses.

**Pull request merged**
